src/extension/csv-with-metadata-header/src/main/java/au/org/emii/geoserver/wfs/response/CsvPivotedFeatureCollectionSource.java

```diff
@@ -0,0 +1,185 @@
+
+
+public class CsvPivotedFeatureCollectionSource implements CsvSource {
+
+    private static final int MAX_PIVOT_COLUMNS = 10000;
+    private final SimpleFeatureIterator featureIterator;
+    private final ArrayList<String> nonPivotColumnNames;
+    private final List<String> pivotColumnNames;
+    private final List<String> pivotSourceColumnNames;
+    private final PivotConfig pivotConfig;
+
+    static Logger LOGGER = org.geotools.util.logging.Logging.getLogger("au.org.emii.geoserver.wfs.response");
+
+    private SimpleFeature nextFeature;
+    private final Catalog catalog;
+    private final ServletContext context;
+
+    public CsvPivotedFeatureCollectionSource(Operation getFeatureOperation, SimpleFeatureCollection featureCollection, Catalog catalog, ServletContext context) {
+        this.featureIterator = featureCollection.features();
+        this.catalog = catalog;
+        this.context = context;
+
+        // Retrieve config params stored in pivot.xml
+
+        PivotConfigFile pivotConfigurationFile = new PivotConfigFile(getDataDirectory(getFeatureOperation));
+        this.pivotConfig = pivotConfigurationFile.getConfig();
+
+        //  Get feature type definition
+        
+        SimpleFeatureType ft = featureCollection.getSchema();
+
+        this.nonPivotColumnNames = new ArrayList<>();
+        pivotSourceColumnNames = new ArrayList<>();
+
+        // Determine pivoted (JSON) and non-pivoted columns
+
+        for (AttributeDescriptor descriptor: ft.getAttributeDescriptors()) {
+            String attributeName = descriptor.getLocalName();
+            String attributeType = descriptor.getUserData().size() > 0 ? descriptor.getUserData().get("org.geotools.jdbc.nativeTypeName").toString() : "";
+
+            if (attributeType.equals(JSON_FIELD_TYPE)) {
+                pivotSourceColumnNames.add(attributeName);
+            } else if (!pivotConfig.getExcludedFields().contains(attributeName)) {
+                this.nonPivotColumnNames.add(attributeName);
+            }
+        }
+
+        // Read the first row to get JSON attr for column names
+
+        if (featureIterator.hasNext()) {
+            nextFeature = featureIterator.next();
+        }
+
+        // Loop through each source column of JSON objects and build up a list of distinct attribute names
+        // alphabetically to then be rendered into columns
+
+        JSONParser parser = new JSONParser();
+        Set<String> foundColumnNames = new HashSet<>();
+        for (String sourceColumnName : pivotSourceColumnNames) {
+            try {
+                FeatureSource featureSource = ((TypeInfoCollectionWrapper.Simple) featureCollection)
+                        .getFeatureTypeInfo().getFeatureSource(null, null);
+                SimpleFeatureCollection features = (SimpleFeatureCollection) featureSource.getFeatures();
+
+                UniqueVisitor visitor = new UniqueVisitor(FF.property(sourceColumnName)) {
```

Comment:
We're after the unique keys - this will return unique json objects.   You'd need to use a postgres query to source the distinct key values with possible performance issues for large datasets - probably OK for WSD though.

Reply:
I guess this will work with the performance proviso - you'll most likely return all records.  But no real need to use a unique visitor.